src/System.Linq/src/System/Linq/Select.SpeedOpt.cs

```diff
@@ -157,6 +157,127 @@ public TResult TryGetLast(out bool found)
             }
         }
 
+        private sealed partial class SelectRangeIterator<TResult> : Iterator<TResult>, IPartition<TResult>
+        {
+            private readonly int _start;
+            private readonly int _end;
+            private readonly Func<int, TResult> _selector;
+
+            public SelectRangeIterator(int start, int end, Func<int, TResult> selector)
+            {
+                Debug.Assert(start < end);
+                Debug.Assert((end - start) <= int.MaxValue);
+                Debug.Assert(selector != null);
+
+                _start = start;
+                _end = end;
+                _selector = selector;
+            }
+
+            public override Iterator<TResult> Clone() =>
+                new SelectRangeIterator<TResult>(_start, _end, _selector);
+
+            public override bool MoveNext()
+            {
+                if (_state < 1 || _state == (_end - _start + 1))
+                {
+                    Dispose();
+                    return false;
+                }
+
+                int index = _state++ - 1;
+                _current = _selector(_start + index);
+                return true;
+            }
+
+            public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> selector) =>
+                new SelectRangeIterator<TResult2>(_start, _end, CombineSelectors(_selector, selector));
+
+            public TResult[] ToArray()
+            {
+                var results = new TResult[_end - _start];
+                int srcIndex = _start;
+                for (int i = 0; i < results.Length; i++)
+                {
+                    results[i] = _selector(srcIndex++);
+                }
+
+                return results;
+            }
+
+            public List<TResult> ToList()
+            {
+                var results = new List<TResult>(_end - _start);
+                for (int i = _start; i != _end; i++)
+                {
+                    results.Add(_selector(i));
```

Comment:
List is guaranteed to be large enough given the initialization. Why not use the same code pattern as the `ToArray()` impl and just use the indexer rather than calling `list.Add()`?

```C#
                int srcIndex = _start;
                for (int i = 0; i < results.Count; i++)
                {
                    results[i] = _selector(srcIndex++);
                }
```

Reply:
Because that won't work :-) Even with an initial capacity that's large enough, you can't index into >= Count

Reply:
Oh, lol. Makes sense. Thanks.